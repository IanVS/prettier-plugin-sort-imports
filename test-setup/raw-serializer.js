'use strict';

import { expect } from 'vitest';

const RAW = Symbol.for('raw');

expect.addSnapshotSerializer({
    print(val) {
        return val[RAW];
    },
    test(val) {
        return (
            val &&
            Object.prototype.hasOwnProperty.call(val, RAW) &&
            typeof val[RAW] === 'string'
        );
    },
});
