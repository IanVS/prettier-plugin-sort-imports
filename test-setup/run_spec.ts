'use strict';

import fs from 'fs';
import { extname } from 'path';

import { format } from 'prettier';
import { expect, test } from 'vitest';

import * as plugin from '../src';

export async function run_spec(dirname, parsers, options) {
    options = Object.assign(
        {
            plugins: [plugin],
            tabWidth: 4,
        },
        options,
    );

    /* instabul ignore if */
    if (!parsers || !parsers.length) {
        throw new Error(`No parsers were specified for ${dirname}`);
    }

    for (const filename of fs.readdirSync(dirname)) {
        const path = dirname + '/' + filename;
        if (
            extname(filename) !== '.snap' &&
            fs.lstatSync(path).isFile() &&
            filename[0] !== '.' &&
            filename !== 'ppsi.spec.ts'
        ) {
            const source = read(path).replace(/\r\n/g, '\n');

            const mergedOptions = Object.assign({}, options, {
                parser: parsers[0],
            });
            test(`${filename} - ${mergedOptions.parser}-verify`, async () => {
                const output = await prettyprint(source, path, mergedOptions);
                expect(
                    raw(source + '~'.repeat(80) + '\n' + output),
                ).toMatchSnapshot(filename);
            });

            for (const parserName of parsers.slice(1)) {
                test(`${filename} - ${parserName}-verify`, async () => {
                    const output = await prettyprint(
                        source,
                        path,
                        mergedOptions,
                    );
                    const verifyOptions = Object.assign(mergedOptions, {
                        parser: parserName,
                    });
                    const verifyOutput = await prettyprint(
                        source,
                        path,
                        verifyOptions,
                    );
                    expect(output).toEqual(verifyOutput);
                });
            }
        }
    }
}

async function prettyprint(src, filename, options) {
    return await format(
        src,
        Object.assign(
            {
                filepath: filename,
            },
            options,
        ),
    );
}

function read(filename) {
    return fs.readFileSync(filename, 'utf8');
}

/**
 * Wraps a string in a marker object that is used by `./raw-serializer.js` to
 * directly print that string in a snapshot without escaping all double quotes.
 * Backticks will still be escaped.
 */
function raw(string) {
    if (typeof string !== 'string') {
        throw new Error('Raw snapshots have to be strings.');
    }
    return { [Symbol.for('raw')]: string };
}
