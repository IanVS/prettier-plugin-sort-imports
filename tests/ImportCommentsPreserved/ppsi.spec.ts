import {run_spec} from '../../test-setup/run_spec';

run_spec(__dirname, ["typescript"], {
    importOrder: ['^@core/(.*)$', '^@server/(.*)', '^@ui/(.*)$', '^[./]'],
    importOrderTypeScriptVersion: "5.0.0",
    importOrderParserPlugins : ['typescript', 'decorators-legacy', 'classProperties']
});
