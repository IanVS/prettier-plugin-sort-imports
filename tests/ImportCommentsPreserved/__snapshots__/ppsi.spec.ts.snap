// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`import-comments-babel-pulling-single-line-bug.ts - typescript-verify > import-comments-babel-pulling-single-line-bug.ts 1`] = `

import B from "b";
// Trailing comment (not supposed to be on same line)
// Bottom of imports single-line comment

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
import B from "b";

// Trailing comment (not supposed to be on same line)
// Bottom of imports single-line comment

`;

exports[`import-comments-directives.ts - typescript-verify > import-comments-directives.ts 1`] = `
/** @jsxImportSource @emotion/react */
"use client";

import * as React from "react";
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/** @jsxImportSource @emotion/react */
"use client";

import * as React from "react";

`;

exports[`import-comments-document-top-test1.ts - typescript-verify > import-comments-document-top-test1.ts 1`] = `
// This is part of multiple-lines at document-top
// Second line followed by a gap

import b from "b";
import a from "a";
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// This is part of multiple-lines at document-top
// Second line followed by a gap

import a from "a";
import b from "b";

`;

exports[`import-comments-document-top-test2.ts - typescript-verify > import-comments-document-top-test2.ts 1`] = `
/**
 * This is part of multiple-lines at document-top
 * Second line of text, comment is followed by a gap
 */

import b from "b";
import a from "a";
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/**
 * This is part of multiple-lines at document-top
 * Second line of text, comment is followed by a gap
 */

import a from "a";
import b from "b";

`;

exports[`import-comments-document-top-test3.ts - typescript-verify > import-comments-document-top-test3.ts 1`] = `
/* eslint-disable jsx-a11y/anchor-is-valid */
/* eslint-disable react/jsx-key */
/* eslint-disable jsx-a11y/alt-text */

import { useState } from "react";
import { Button, Menu, MenuButton, MenuItem, MenuList } from "@chakra-ui/react";
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/* eslint-disable jsx-a11y/anchor-is-valid */
/* eslint-disable react/jsx-key */
/* eslint-disable jsx-a11y/alt-text */

import { Button, Menu, MenuButton, MenuItem, MenuList } from "@chakra-ui/react";
import { useState } from "react";

`;

exports[`import-comments-outer-comments-block-comments.ts - typescript-verify > import-comments-outer-comments-block-comments.ts 1`] = `
// Loose leading comment before imports (should not be dragged down with B)

import C from "c" /* trailing comment that spans
                     onto the following line */

/* Leading comment before B */
import B from "b"; /* Trailing comment on same-line as B */
/* Trailing comment on first line after B (this is treated as a leading comment for next import A!) */


/* Leading comment before A */
import A from "a"; /* Trailing comment on same-line as A */
/* Trailing comment on first line after A (this is treated as a "bottom-of-imports comment) */

// Loose trailing comment after imports (should not be dragged up with A)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Loose leading comment before imports (should not be dragged down with B)

/* Trailing comment on first line after B (this is treated as a leading comment for next import A!) */

/* Leading comment before A */
import A from "a"; /* Trailing comment on same-line as A */
/* Leading comment before B */
import B from "b"; /* Trailing comment on same-line as B */
import C from "c"; /* trailing comment that spans
                      onto the following line */

/* Trailing comment on first line after A (this is treated as a "bottom-of-imports comment) */

// Loose trailing comment after imports (should not be dragged up with A)

`;

exports[`import-comments-outer-comments-single-line.ts - typescript-verify > import-comments-outer-comments-single-line.ts 1`] = `
// Loose leading comment before imports (should not be dragged down with B)

// Leading comment before B
import B from "b"; // Trailing comment on same-line as B
// Trailing comment on first line after B (this is treated as a leading comment for next import A!)

// Leading comment before A
import A from "a"; // Trailing comment on same-line as A
// Trailing comment on first line after A (this is treated as a "bottom-of-imports comment)

// Loose trailing comment after imports (should not be dragged up with A)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Loose leading comment before imports (should not be dragged down with B)

// Trailing comment on first line after B (this is treated as a leading comment for next import A!)

// Leading comment before A
import A from "a"; // Trailing comment on same-line as A
// Leading comment before B
import B from "b"; // Trailing comment on same-line as B

// Trailing comment on first line after A (this is treated as a "bottom-of-imports comment)

// Loose trailing comment after imports (should not be dragged up with A)

`;

exports[`import-comments-specifier-leading-and-trailing.ts - typescript-verify > import-comments-specifier-leading-and-trailing.ts 1`] = `
// Top-of-file

import {
    // j3 leading single-line-comment 1
    // j3 leading single-line-comment 2
    j3, // j3 same-line single-line-comment
    j2, // j2 same-line single-line-comment
} from "j";

import {
    /** i3 leading comment
      */ i3,
    i2, /* i2 same-line block */
    i4, /* i4 multi-line
        block */
    i1, /* i1 same-line block */
} from "i";

import {
    h4, // h4 same-line single-line-comment
    // h3 leading single-line-comment
    h3, // h3 same-line single-line-comment
    h2, // h2 same-line single-line-comment
    h1, // h1 same-line single-line-comment
} from "h";

// Rest of file below here
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// Top-of-file

import {
    h1, // h1 same-line single-line-comment
    h2, // h2 same-line single-line-comment
    // h3 leading single-line-comment
    h3, // h3 same-line single-line-comment
    h4, // h4 same-line single-line-comment
} from "h";
import {
    i1 /* i1 same-line block */,
    i2 /* i2 same-line block */,
    /** i3 leading comment
     */
    i3,
    i4 /* i4 multi-line
       block */,
} from "i";
import {
    j2, // j2 same-line single-line-comment
    // j3 leading single-line-comment 1
    // j3 leading single-line-comment 2
    j3, // j3 same-line single-line-comment
} from "j";

// Rest of file below here

`;

exports[`import-comments-trailing-specifier-comment-lost.ts - typescript-verify > import-comments-trailing-specifier-comment-lost.ts 1`] = `
import {
    b3,
    a3,
    /* @ts-expect-error*/
} from "c";

import {
    b4,
    a4,
    // @ts-expect-error
} from "c2";

import {
    b2,a2,
    // @ts-expect-error
} from "b";

import {
    b1,
    a1,

    // @ts-expect-error
} from "a";
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
import {
    a1,
    b1,
    // @ts-expect-error
} from "a";
import {
    a2,
    b2,
    // @ts-expect-error
} from "b";
import { a3, b3 } from /* @ts-expect-error*/ "c";
import {
    a4,
    b4,
    // @ts-expect-error
} from "c2";

`;
